### PR TITLE
test(core): Flush promises on telemetry tests

### PR DESCRIPTION
**Pull request by ivov**
@freyamade No context on the feature so please adjust as needed. All passing except last three assertions in second test still pending.

For reference, telemetry tests pass on node 14.16.1, 16.18 and 16.9, but fail on 16.13.1. Officially we recommend 16.9 so the issue yday was a false flag.

### CODE REVIEW

File: packages/cli/test/unit/Telemetry.test.ts [hunk 1]
```diff
@@ -281,23 +281,34 @@ describe('Telemetry', () => {
 			pulseSpy.mockClear();
 		});
 
-		xtest('should trigger pulse in intervals', () => {
+		test('should trigger pulse in intervals', async () => {
 			expect(pulseSpy).toBeCalledTimes(0);
 
 			jest.advanceTimersToNextTimer();
+			await Promise.resolve();
```
Comment (netroy) [inline packages/cli/test/unit/Telemetry.test.ts:288]: maybe we could extract this into a helper, and use `setImmediate` for resolving the promises, to give any pending promises some time to finish?
something like this
```typescript
const flushPromises = async () => new Promise((resolve) => setImmediate(resolve));
```

```typescript
await flushPromises();
```

Comment (ivov) [reply packages/cli/test/unit/Telemetry.test.ts]: `const flushPromises = async () => new Promise(setImmediate);` and `const flushPromises = () => new Promise(setImmediate);` trigger timeouts (not sure why), but `export const flushPromises = () => Promise.resolve(setImmediate);` works fine so went with that.

Comment (netroy) [reply packages/cli/test/unit/Telemetry.test.ts]: `const flushPromises = () => Promise.resolve(setImmediate);` creates a new resolved promise with `setImmediate` as the value. it's not actually using the `setImmediate` to let any dangling promises in the current event-loop from finishing.

Comment (ivov) [reply packages/cli/test/unit/Telemetry.test.ts]: Got it, I see now, I'll dig into these timeouts then. Also something about this setup is OOMing the CI.

Comment (ivov) [reply packages/cli/test/unit/Telemetry.test.ts]: @freyamade Adi might have traced this back to the PH SDK:

> btw, the timeout issues are actually expected. we most certainly have handling promises that aren't being resolved. the flushPromise thing will only work if all promises resolve in a timely manner.
adding a simple await Promise.resolve() isn't going to help with promises that never resolve :confused: (edited) 
in this case, looks like the dangling promise is from the posthog sdk.

Up to you if we want to spend more time to unskip these two tests at this stage.

Comment (ivov) [reply packages/cli/test/unit/Telemetry.test.ts]: Related #4930

**Comment by krynble** [comment]
I've cherry picked the commits from here to the main feature branch and I'll be closing this - thanks a lot @ivov and @netroy

**Pull request closed without merging**